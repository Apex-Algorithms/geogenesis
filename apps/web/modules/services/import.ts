import { parse as parseCSV } from 'papaparse';
import { Triple, Value } from '../types';
import { BUILTIN_ENTITY_IDS, createEntityId, createTripleWithId, isValidEntityId } from './create-id';

function readFileAsText(file: File) {
  return new Promise<string>(resolve => {
    const reader = new FileReader();
    reader.addEventListener('load', event => {
      const data = event.target!.result;
      resolve(data as string);
    });
    reader.readAsText(file);
  });
}

type CreateUuid = (value: string) => string;

function readCSV(csv: string, createId: CreateUuid = createEntityId): Triple[] {
  const results = parseCSV<string[]>(csv);

  const rows = results.data.slice(1);

  if (rows.some(row => row.length < 3)) {
    throw new Error('Each row must have 3 cells');
  }

  // Create a collection of all known entity ids
  const entityIds = new Set([
    ...BUILTIN_ENTITY_IDS,
    ...rows.flatMap(([entityId, attributeId]) => [entityId, attributeId]),
  ]);

  // Ensure entity ids are either valid uuids or builtins
  const entityIdMap = Object.fromEntries([...entityIds].map(id => [id, isValidEntityId(id) ? id : createId(id)]));

  // Create triples, attempting to detect entity references
  const triples = rows.map((row): Triple => {
    const [entityId, attributeId, value] = row;

    const mappedEntityId = entityIdMap[entityId];
    const mappedAttributeId = entityIdMap[attributeId];
    const mappedValue: Value =
      value in entityIdMap ? { type: 'entity', value: entityIdMap[value] } : { type: 'string', value };

    return createTripleWithId(mappedEntityId, mappedAttributeId, mappedValue);
  });

  return triples;
}

export async function importCSVFile(file: File, createId: CreateUuid = createEntityId): Promise<Triple[]> {
  const csv = await readFileAsText(file);
  return readCSV(csv, createId);
}
