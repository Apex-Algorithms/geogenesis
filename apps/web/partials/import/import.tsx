'use client';

import { Space } from '~/core/io/dto/spaces';

import { Generate } from './generate';
import { Publish } from './publish';

type ImportProps = {
  spaceId: string;
  space: Space;
};

export const Import = ({ spaceId, space }: ImportProps) => {
  return (
    <>
      <Generate spaceId={spaceId} space={space} />
      <Publish spaceId={spaceId} space={space} />
    </>
  );
};
