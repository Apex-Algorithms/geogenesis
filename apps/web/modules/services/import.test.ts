import { readFileSync } from 'fs';
import path from 'path';
import { v4 as uuid } from 'uuid';
import { describe, expect, it } from 'vitest';
import { importCSVFile } from './import';

function readMockFile(filename: string) {
  const simple = readFileSync(path.join(__dirname, 'mocks', filename), 'utf8');
  return new File([simple], filename);
}

function deterministicUuid(value: string) {
  const random = value
    .slice(0, 16)
    .padEnd(16, '0')
    .split('')
    .map(c => c.charCodeAt(0));

  return uuid({ random });
}

describe('CSV Import', () => {
  it('imports csv', async () => {
    const file = readMockFile('simple.csv');
    const triples = await importCSVFile(file, deterministicUuid);

    expect(triples).toMatchSnapshot();
  });
});
