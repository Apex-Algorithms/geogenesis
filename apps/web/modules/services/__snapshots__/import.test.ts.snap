// Vitest Snapshot v1

exports[`CSV Import > imports csv 1`] = `
[
  {
    "attributeId": "name",
    "entityId": "31303030-3030-4030-b030-303030303030",
    "id": "31303030-3030-4030-b030-303030303030:name:s~Devin",
    "value": {
      "type": "string",
      "value": "Devin",
    },
  },
  {
    "attributeId": "type",
    "entityId": "31303030-3030-4030-b030-303030303030",
    "id": "31303030-3030-4030-b030-303030303030:type:e~70657273-6f6e-4030-b030-303030303030",
    "value": {
      "type": "entity",
      "value": "70657273-6f6e-4030-b030-303030303030",
    },
  },
  {
    "attributeId": "name",
    "entityId": "32303030-3030-4030-b030-303030303030",
    "id": "32303030-3030-4030-b030-303030303030:name:s~Sam",
    "value": {
      "type": "string",
      "value": "Sam",
    },
  },
  {
    "attributeId": "type",
    "entityId": "32303030-3030-4030-b030-303030303030",
    "id": "32303030-3030-4030-b030-303030303030:type:e~70657273-6f6e-4030-b030-303030303030",
    "value": {
      "type": "entity",
      "value": "70657273-6f6e-4030-b030-303030303030",
    },
  },
  {
    "attributeId": "type",
    "entityId": "70657273-6f6e-4030-b030-303030303030",
    "id": "70657273-6f6e-4030-b030-303030303030:type:e~type",
    "value": {
      "type": "entity",
      "value": "type",
    },
  },
  {
    "attributeId": "name",
    "entityId": "70657273-6f6e-4030-b030-303030303030",
    "id": "70657273-6f6e-4030-b030-303030303030:name:s~Person",
    "value": {
      "type": "string",
      "value": "Person",
    },
  },
]
`;
